swmmio/core.py

```diff
@@ -1508,6 +1512,98 @@ def timeseries(self, df):
         self._timeseries_df = df
 
+    @property
+    def patterns(self):
+        """
+        Get/set patterns section of the model
+        
+        :return: dataframe of patterns
+        
+        >>> from swmmio.examples import pump_control
+        >>> # NOTE, only the first 5 columns are shown in the following example
+        >>> pump_control.inp.patterns.iloc[:,0:5]  #doctest: +NORMALIZE_WHITESPACE
+                Type  Factor1  Factor2  Factor3  Factor4
+        Name                                            
+        DWF   HOURLY   0.0151  0.01373  0.01812  0.01098
+        """
+        
+        if self._patterns_df is not None:
+            return self._patterns_df
+        self._patterns_df = dataframe_from_inp(self.path, '[PATTERNS]')
+
+        if self._patterns_df.shape[0] > 0:
+            # reformat, 1 row per pattern
+            pattern_entry_list = []
+            for name, pattern in self._patterns_df.groupby('Name'):
+                pattern_entry = {}
+                pattern_entry['Name'] = name
+                pattern_entry['Type'] = pattern['Type'].iloc[0]
+                if pattern.shape[0] > 1:
+                    # shift pattern values to the right
+                    pattern.iloc[1::, 1::] = pattern.iloc[1::, 0:-1].values
+                    pattern['Factors'] = pattern['Factors'].astype(float)
+                values = pattern.iloc[:, 1:].values.flatten()
+                for i in range(len(values)):
+                    pattern_entry['Factor'+str(i+1)] = values[i]
+                pattern_entry_list.append(pattern_entry)
+
+            self._patterns_df = pd.DataFrame(pattern_entry_list)
+            self._patterns_df.set_index('Name', inplace=True)
+
+        return self._patterns_df
+
+    @patterns.setter
+    def patterns(self, df):
+        """Set inp.patterns DataFrame."""
+        self._patterns_df = df
+
+    @property
+    def controls(self):
+        """
+        Get/set controls section of the model
+        
+        :return: dataframe of controls
+        
+        >>> from swmmio.examples import pump_control
+        >>> pump_control.inp.controls  #doctest: +NORMALIZE_WHITESPACE
+                                                                            Control
+        Name                                                                       
+        RULE PUMP1A  IF NODE SU1 DEPTH >= 4 THEN PUMP PUMP1 status = ON PRIORITY 1 
+        RULE PUMP1B  IF NODE SU1 DEPTH < 1 THEN PUMP PUMP1 status = OFF PRIORITY 1 
+        """
+        
+        if self._controls_df is None:
+            self._controls_df = dataframe_from_inp(self.path, "[CONTROLS]")
+        
+        if self._controls_df.shape[0] > 0:
+            # reformat, 1 row per control
+            control_entry_list = []
+            control_entry = {}
+            controls = self._controls_df['[CONTROLS]']
+            # make sure the first entry starts with RULE
+            assert controls[0][0:5] == "RULE "
```

Comment:
I think it would best to raise an exception here if this assertion fails. But I'm fine with capturing that in a future PR.